File: WdtcCore/src/main/java/org/wdt/wdtc/core/download/fabric/FabricInstallTask.java
package org.wdt.wdtc.core.download.fabric;

import com.google.gson.JsonArray;
import com.google.gson.JsonObject;
import org.wdt.utils.dependency.DependencyDownload;
import org.wdt.utils.gson.JsonUtils;
import org.wdt.wdtc.core.download.infterface.DownloadSourceInterface;
import org.wdt.wdtc.core.download.infterface.InstallTaskInterface;
import org.wdt.wdtc.core.download.infterface.VersionJsonObjectInterface;
import org.wdt.wdtc.core.game.GameVersionJsonObject;
import org.wdt.wdtc.core.game.Launcher;
import org.wdt.wdtc.core.game.LibraryObject;
import org.wdt.wdtc.core.manger.DownloadSourceManger;
import org.wdt.wdtc.core.utils.DownloadUtils;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class FabricInstallTask extends FabricDonwloadInfo implements InstallTaskInterface {
  private final DownloadSourceInterface source;


  public FabricInstallTask(Launcher launcher, String FabricVersionNumber) {
    super(launcher, FabricVersionNumber);
    this.source = DownloadSourceManger.getDownloadSource();
  }

  public FabricInstallTask(Launcher launcher, VersionJsonObjectInterface versionJsonObjectInterface) {
    this(launcher, versionJsonObjectInterface.getVersionNumber());
  }

  @Override
  public void overwriteVersionJson() throws IOException {
    GameVersionJsonObject VersionJsonObject = launcher.getGameVersionJsonObject();
    List<LibraryObject> libraryObjectList = VersionJsonObject.getLibraries();
    JsonObject FabricVersionJsonObject = getFabricVersionJsonObject();
    JsonArray FabricLibraryList = FabricVersionJsonObject.getAsJsonArray("libraries");
    for (int i = 0; i < FabricLibraryList.size(); i++) {
      JsonObject object = FabricLibraryList.get(i).getAsJsonObject();
      DependencyDownload dependency = new DependencyDownload(object.get("name").getAsString());
      dependency.setDefaultUrl(source.getFabricLibraryUrl());
      dependency.setDownloadPath(launcher.getGameLibraryDirectory());
      libraryObjectList.add(LibraryObject.getLibraryObject(dependency, object.get("url").getAsString()));
    }
    VersionJsonObject.setLibraries(libraryObjectList);
    VersionJsonObject.setMainClass(FabricVersionJsonObject.get("mainClass").getAsString());
    GameVersionJsonObject.Arguments arguments = VersionJsonObject.getArguments();
    JsonObject Arguments = FabricVersionJsonObject.get("arguments").getAsJsonObject();
    JsonArray JvmList = arguments.getJvmList();
    JvmList.addAll(Arguments.getAsJsonArray("jvm"));
    arguments.setJvmList(JvmList);
    JsonArray GameList = arguments.getGameList();
    GameList.addAll(Arguments.getAsJsonArray("game"));
    arguments.setGameList(GameList);
    VersionJsonObject.setArguments(arguments);
    VersionJsonObject.setId(launcher.getVersionNumber() + "-fabric-" + FabricVersionNumber);
    launcher.putToVersionJson(VersionJsonObject);
  }


  @Override
  public void writeVersionJsonPatches() throws IOException {
    GameVersionJsonObject Object = launcher.getGameVersionJsonObject();
    List<JsonObject> ObjectList = new ArrayList<>();
    ObjectList.add(JsonUtils.getJsonObject(launcher.getVersionJson()));
    ObjectList.add(JsonUtils.getJsonObject(getFabricVersionJson()));
    Object.setJsonObject(ObjectList);
    launcher.putToVersionJson(Object);
  }

  @Override
  public void afterDownloadTask() throws IOException {
    if (isAPIDownloadTaskNoNull()) {
      getApiDownloadTask().startDownloadFabricAPI();
    }
  }

  @Override
  public void beforInstallTask() {
    DownloadUtils.StartDownloadTask(String.format(getFabricVersionFileUrl(), launcher.getVersionNumber(), FabricVersionNumber), getFabricVersionJson());
  }
}

File: WdtcCore/src/main/kotlin/org/wdt/wdtc/core/download/fabric/FabricInstallTask.kt
package org.wdt.wdtc.core.download.fabric

import org.wdt.utils.gson.*
import org.wdt.wdtc.core.download.infterface.ModInstallTaskInterface
import org.wdt.wdtc.core.download.infterface.VersionsJsonObjectInterface
import org.wdt.wdtc.core.game.*
import org.wdt.wdtc.core.manger.downloadSource
import org.wdt.wdtc.core.utils.ckeckIsNull
import org.wdt.wdtc.core.utils.startDownloadTask
import org.wdt.wdtc.core.utils.toURL
import java.io.IOException

class FabricInstallTask(version: Version, fabricVersionNumber: String) :
  FabricDonwloadInfo(version, fabricVersionNumber), ModInstallTaskInterface {

  constructor(version: Version, versionsJsonObjectInterface: VersionsJsonObjectInterface) : this(
    version,
    versionsJsonObjectInterface.versionNumber.ckeckIsNull()
  )

  @Throws(IOException::class)
  override fun overwriteVersionJson() {
    val gameVersionJsonObject = version.gameVersionJsonObject
    val libraryObjectList = gameVersionJsonObject.libraries
    val fabricVersionJsonObject = fabricVersionJsonObject
    val fabricLibraryList = fabricVersionJsonObject.getJsonArray("libraries")
    for (i in 0 until fabricLibraryList.size()) {
      val libraryObject = fabricLibraryList.getJsonObject(i)
      val dependency = GameRuntimeDependency(libraryObject.getString("name"))
      dependency.libraryRepositoriesUrl = downloadSource.fabricLibraryUrl.toURL()
      dependency.libraryDirectory = version.gameLibraryDirectory
      libraryObjectList.add(LibraryObject.getLibraryObject(dependency, libraryObject.getString("url").toURL()))
    }
    gameVersionJsonObject.libraries = libraryObjectList
    gameVersionJsonObject.mainClass = fabricVersionJsonObject.getString("mainClass")
    val gameVersionJsonArguments = gameVersionJsonObject.arguments
    val fabricVersionJsonArguments = fabricVersionJsonObject.getJsonObject("arguments")
    val gameVersionJsonJvmList = gameVersionJsonArguments.jvmList!!
    gameVersionJsonJvmList.addAll(fabricVersionJsonArguments.getJsonArray("jvm"))
    gameVersionJsonArguments.jvmList = gameVersionJsonJvmList
    val gameVersionJsonGameList = gameVersionJsonArguments.gameList!!
    gameVersionJsonGameList.addAll(fabricVersionJsonArguments.getJsonArray("game"))
    gameVersionJsonArguments.gameList = gameVersionJsonGameList
    gameVersionJsonObject.arguments = gameVersionJsonArguments
    gameVersionJsonObject.id = "${version.versionNumber}-fabric-$modVersion"
    version.run {
      gameVersionJsonObject.putToVersionJson()
    }
  }

  @Throws(IOException::class)
  override fun writeVersionJsonPatches() {
    version.run {
      gameVersionJsonObject.apply {
        patches = mutableListOf(
          version.versionJson.readFileToJsonObject(),
          fabricVersionJson.readFileToJsonObject()
        )
      }.putToVersionJson()
    }
  }

  @Throws(IOException::class)
  override fun afterDownloadTask() {
    if (isAPIDownloadTaskNoNull) {
      apiDownloadTask.ckeckIsNull().startDownloadFabricAPI()
    }
  }

  override fun beforInstallTask() {
    startDownloadTask(
      fabricVersionFileUrl.format(version.versionNumber, modVersion),
      fabricVersionJson
    )
  }
}